### PR TITLE
Create module for SWC-124: Arbitrary Write

### LINKED ISSUE #861: Analysis Module: SWC-124 - Write to Arbitrary Storage Location
Create an analysis module for SWC-124 - Write to Arbitrary Storage Location. SWC link: <https://smartcontractsecurity.github.io/SWC-registry/docs/SWC-124>

This can be done by adding pre-hooks at `sstore` instructions and checking if it is possible to constrain the location argument to an arbitrary value.

### PR DESCRIPTION
**Pull request by nbanmp**
closes #861 

After putting this module together, I realized that mythril adds unconstrained bitvecs to the stack fairly often when we don't know the correct value to add to the stack is. This might result in way too many false positives, and as such, we might not want to merge it.

### CODE REVIEW

File: mythril/analysis/modules/arbitrary_write.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    """This module checks if an arbitrary storage location in the contact can be written to by anyone."""
+
+    def __init__(self):
+        super().__init__(
+            name="Write to Arbitrary Storage",
+            swc_id=WRITE_TO_ARBITRARY_STORAGE,
+            description=DESCRIPTION,
+            entrypoint="callback",
+            pre_hooks=["SSTORE"],
+        )
```
Comment (norhh) [inline mythril/analysis/modules/arbitrary_write.py]: Can you add a `-> None`  in this part

File: mythril/analysis/modules/arbitrary_write.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        self._cache_address = {}
+
+    def reset_module(self):
```
Comment (norhh) [inline mythril/analysis/modules/arbitrary_write.py]: this returns None. `-> None`

File: mythril/analysis/modules/arbitrary_write.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        Resets the module
+        :return:
+        """
+        super().reset_module()
+        self._cache_address = {}
+
+    def execute(self, state: GlobalState):
```
Comment (norhh) [inline mythril/analysis/modules/arbitrary_write.py]: Returns a list of issues

File: mythril/analysis/modules/arbitrary_write.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        """
+
+        :param state:
+        :return:
+        """
+        self._issues.extend(self._analyze_state(state))
+        return self.issues
+
+    def _analyze_state(self, state):
```
Comment (norhh) [inline mythril/analysis/modules/arbitrary_write.py]: returns a list of issues.

**Comment by JoranHonig** [comment]
@nbanmp Does this module perhaps give a ton of false positives when there is a mapping?

**Comment by nbanmp** [comment]
@JoranHoing From my relatively small amount of testing so far, yes, it looks like the issues are often returned when there is a mapping.

**Comment by JoranHonig** [comment]
:upside_down_face: I think this pr will have to wait until we have made a neat implementation for storage & uninterpreted functions for hashes

File: mythril/analysis/modules/arbitrary_write.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    def __init__(self):
+        super().__init__(
+            name="Write to Arbitrary Storage",
+            swc_id=WRITE_TO_ARBITRARY_STORAGE,
+            description=DESCRIPTION,
+            entrypoint="callback",
+            pre_hooks=["SSTORE"],
+        )
+        self._cache_address = {}
```
Comment (norhh) [inline mythril/analysis/modules/arbitrary_write.py]: This should also be typehinted as `Dict[int, bool]` or something like that.

**Comment by nbanmp** [comment]
Closing until storage implementation is refactored.

**Pull request closed without merging**